File: day-1/day-1.py
from functools import reduce

def get_input() -> list[str]:
    with open("day-1/input.txt", 'r') as file:
        return file.readlines()

def solve() -> int:
    data = get_input()
    left, right = reduce(lambda prev, curr : (prev[0] + [curr[0]], prev[1] + [curr[1]]), map(lambda line : tuple(map(int, line.split("   "))), data), [[],[]]) 
    return sum(map(lambda mins : abs(mins[0] - mins[1]), zip(sorted(left), sorted(right))))

if __name__ == "__main__":
    print(solve())

File: day-1/day-1.hs
import Data.List
import Text.Printf

getInput :: IO [[Int]]
getInput = do
    lists <- readFile "day-1/input.txt"
    let unzipped = map (map read . words) $ lines lists :: [[Int]]
    return . foldl (\(left:right:_) (first:second:_) -> [left ++ [first], right ++ [second]]) [[],[]] $ unzipped

partOne :: IO Int
partOne = do
    left:right:_ <- getInput
    return . sum . map abs $ zipWith (-) (sort left) (sort right)

count :: Eq a => a -> [a] -> Int
count elem = length . filter (== elem)

partTwo :: IO Int
partTwo = do
    left:right:_ <- getInput
    return . sum . zipWith (*) left . map (`count` right) $ left  

main :: IO ()
main = do
    partOneAnswer <- partOne
    partTwoAnswer <- partTwo
    printf "Part One: %d\nPart Two: %d\n" partOneAnswer partTwoAnswer
    